### PR TITLE
Nataliya - Time

### PR DESCRIPTION
**Pull request by npogodina**
# Assignment Submission: Video Store API

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Reflection

Prompt | Response
--- | ---
Explain how you came up with the initial design of your ERD, based on the seed data and reading through the API endpoints | The seeds contained the available information for customers and videos which I turned into two Models (Customer and Video) with corresponding attributes. The API documentation gave insight into the renting process which allowed me to design a Rental model.
What would be the Big-O time complexity of your `/customers` & `/videos` endpoints? What does the time complexity depend on? Explain your reasoning. | The index route will be O(n) because the size of the displayed data will directly depend on the table size in the database. The show route will be O(1) because it's a look-up with id (primary key). And for the create route, I want to say O(1) because it merely inserts one entry into the table, but it probably makes sense to call it O(n) as well because it somewhat depends on the size of the input (what if the video has a very long overview? :))
What is the Big-O time complexity of the `POST /rentals/check-in` endpoint? What does the time complexity depend on? Explain your reasoning.  | I think check-in route is O(n), because it has to find the rental which matches the requirements (provided customer ID and video ID). Since rental ID is not provided, it's not a fast look-up.
Describe a specific set of positive and negative test cases you implemented for a model. | For example, I have a validation rule for :total_inventory attribute of Video model - it is required and it should be an integer greater than 0. I have a positive test case for when it is a valid number, and multiple negative tests - when it is not an integer, but a string, when it is a negative number and when it is zero.
Describe a specific set of positive and negative test cases you implemented for a controller. | When testing an index route for videos, I expect the response to be 200 success and in JSON format. Specifically, I expect the response to be an array of hashes with information about videos. And I also test for a negative (edge) case when there are no videos in the database.
Broadly, describe how an API should respond when it handles a request with invalid/erroneous parameters. | It renders JSON with errors field that displays the error message (what went wrong). It also returns a corresponding response status (not fount or bad request).
Describe one of your custom model methods and why you chose to wrap that functionality into a method. | I have a default values method for Rental model which sets the due date to 7 days in advance from today's date. I call this method before saving the entry so that it automatically fills in the due date.

### CODE REVIEW

File: app/controllers/rentals_controller.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      rental.video.available_inventory -= 1
+      rental.video.save
+      
+      render json: {
+        customer_id: rental.customer_id,
+        video_id: rental.video_id,
+        due_date: rental.due_date,
+        videos_checked_out_count: rental.customer.videos_checked_out_count,
+        available_inventory: rental.video.available_inventory
+      }, status: :ok
```
Comment (kaidamasaki) [inline app/controllers/rentals_controller.rb:25]: Most of this logic should have been encapsulated into a model method to handle checking out.  

```suggestion
    if video.available_inventory == 0
      render json: { errors: ["No available copies of the video available"] }, status: :bad_request
      return
    end
    
    rental = Rental.new(customer: @customer, video: @video)
    
    if rental.check_out
      render json: {
        customer_id: rental.customer_id,
        video_id: rental.video_id,
        due_date: rental.due_date,
        videos_checked_out_count: rental.customer.videos_checked_out_count,
        available_inventory: rental.video.available_inventory
      }, status: :ok
```

```ruby
class Rental < ApplicationRecord
  # Relationships, validations, etc...

  def check_out(customer, video)
    if rental.save
      rental.customer.videos_checked_out_count += 1
      rental.customer.save
      rental.video.available_inventory -= 1
      rental.video.save
      
      return rental
    else
      return nil
    end
  end
end
```

File: app/controllers/rentals_controller.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  def check_in
+    rental = Rental.find_by(video_id: params[:video_id], customer_id: @customer.id)
+    
+    if rental
+      rental.destroy
+      rental.customer.videos_checked_out_count -= 1
+      rental.customer.save
+      rental.video.available_inventory += 1
+      rental.video.save
```
Comment (kaidamasaki) [inline app/controllers/rentals_controller.rb:42]: Similar to above this should probably live in a `rental.check_in` method.

```ruby
class Rental < ApplicationRecord
  # ...
  def check_in
    customer.videos_checked_out_count -= 1
    customer.save
    video.available_inventory += 1
    video.save

    self.destroy
  end
end
```

File: app/controllers/rentals_controller.rb (hunk 1 above)
Comment (kaidamasaki) [reply app/controllers/rentals_controller.rb:25]: In a production application you would also probably want to use a [transaction](https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html) for safety when updating multiple tables like this, though we haven't taught you those (and don't as part of the curriculum due to a lack of time).

File: app/controllers/rentals_controller.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      render json: { errors: ["Not Found"] }, status: :not_found
+    end
+  end
+
+  def require_video
+    @video = Video.find_by(id: params[:video_id])
+    if @video.nil?
+      render json: { errors: ["Not Found"] }, status: :not_found
+    end
+  end
```
Comment (kaidamasaki) [inline app/controllers/rentals_controller.rb:72]: Excellent controller filters!

File: app/models/rental.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  belongs_to :customer
+
+  validates :video_id, presence: true, uniqueness: {scope: :customer_id}
+  validates :customer_id, presence: true
+
+  before_save :default_values
+  def default_values
+    self.active = true if self.active.nil? 
+    self.due_date = Date.today + 7.day if self.due_date.nil?
+  end
```
Comment (kaidamasaki) [inline app/models/rental.rb:12]: Good use of `before_safe`!  I especially like that your `default_values` method checks to make sure not to overwrite existing values.

File: test/controllers/rentals_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        videos_checked_out_count: 0
+      )
+
+      Video.create(
+        title: "Cinderella",
+        overview: "After her father unexpectedly dies, young Ella (Lily James) finds herself at the mercy of her cruel stepmother (Cate Blanchett) and stepsisters, who reduce her to scullery maid. Despite her circumstances, she refuses to despair.",
+        release_date: "2015-03-06",
+        total_inventory: 5,
+        available_inventory: 5
+      )
```
Comment (kaidamasaki) [inline test/controllers/rentals_controller_test.rb:33]: This probably should have been part of your fixtures.

File: test/controllers/rentals_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+      @customer = Customer.find_by(name: "Evelynn")
+      @video = Video.find_by(title: "Cinderella")
```
Comment (kaidamasaki) [inline test/controllers/rentals_controller_test.rb:36]: You do additional `find_by`s for these later, which you don't need to do since they already exist in instance variables.

File: test/controllers/rentals_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      expect(body["available_inventory"]).must_equal video.available_inventory
+    end
+
+    it "increase the customer's videos_checked_out_count by one" do
+      expect(@customer.videos_checked_out_count).must_equal 0
+      
+      post rentals_path, params: @rental_data
+      customer = Customer.find_by(name: "Evelynn")
+      expect(customer.videos_checked_out_count).must_equal 1
+    end
```
Comment (kaidamasaki) [inline test/controllers/rentals_controller_test.rb:73]: You could make this cleaner by using `must_differ`.

Also, it's a good idea to assert on the response code here as well.

```suggestion
    it "increase the customer's videos_checked_out_count by one" do
      expect do
        post rentals_path, params: @rental_data
      end.must_differ "@customer.refresh.videos_checked_out_count", 1
      
      must_respond_with :ok
    end
```

File: config/routes.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :videos, only: [:index, :show, :create]
+  resources :customers, only: [:index]
+  resources :rentals, only: [:create, :destroy]
```
Comment (kaidamasaki) [inline config/routes.rb:5]: You didn't need these routes since you only interact with rentals using the custom routes below.

**Comment by kaidamasaki** [comment]
# Video Store API



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Practices git with at least 10 small commits and meaningful commit messages | ✔️
| Understands the importance of how APIs handle invalid/erroneous data in their reflection questions | ✔️
| Practices Rails best practices and well-designed separation, and encapsulates business logic around check-out and check-in in models | Check-in/out logic in controller.
| Uses controller tests to ensure quality code for every route, and checks for response status and content, with positive cases and negative cases | ✔️
| Uses controller tests to ensure correctness for `check out` and `check in` requirement, and that checked out counts and inventories appropriately change | ✔️

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| All provided smoke tests for Wave 1 pass | ✔️
| All provided smoke tests for Wave 2 pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 3+ in Code Review && 2 in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 2+ in Code Review && 1+ in Functional Requirements, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-1 in Code Review or 0 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 


### Additional Feedback


#### Comprehension Questions



I'm just expanding on these because I know that you like detailed explanations.



> The show route will be O(1) because it's a look-up with id (primary key).


Looking up on the primary key in a database is generally an O(log(n)) operation since database tables are stored as trees (and are typically organized by primary keys).  However, the base of that log tends to be _very_ large because they use a structure called a [B-Tree](https://en.wikipedia.org/wiki/B-tree).  (Feel free to ask me more about B-Trees if the Wikipedia article is confusing.)


> I think check-in route is O(n), because it has to find the rental which matches the requirements (provided customer ID and video ID). Since rental ID is not provided, it's not a fast look-up.


This is actually also a quite fast operation.  Because you added indexes to your foreign keys the database actually created one additional B-Tree per reference which means looking up against each of these is a O(log(n)) operation which means that the total complexity of check-in is O(log(n)).


(A nice thing about databases is that you don't have to tell them to use indexes, they just do automatically when you reference columns.)




## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized |

**Review by kaidamasaki — COMMENTED** [review]
Great job!  

Overall your code was quite clean and clear.  There were a few little things I'd refresh on for your next project.  It's probably worth taking another look at fixtures and model methods.

Your controllers were clearly defined and I appreciated how you split out your logic and especially how thorough your tests were.
